File: tests/fixtures/benchmark_project/frontend/types/order.ts
/**
 * Order-related type definitions.
 */

import { User, UserID } from "./user";
import { Product, ProductID } from "./product";

export enum OrderStatus {
    Pending = "pending",
    Confirmed = "confirmed",
    Shipped = "shipped",
    Delivered = "delivered",
    Cancelled = "cancelled",
}

export interface OrderItem {
    product: Product;
    quantity: number;
    subtotal: number;
}

export interface Order {
    id: number;
    user: User;
    items: OrderItem[];
    total: number;
    status: OrderStatus;
    createdAt: string;
    notes?: string;
}

export type OrderID = number;

export interface CreateOrderRequest {
    userId: UserID;
}

export interface AddItemRequest {
    productId: ProductID;
    quantity?: number;
}

export function calculateOrderTotal(items: OrderItem[]): number {
    return items.reduce((sum, item) => sum + item.subtotal, 0);
}

export function isOrderEditable(order: Order): boolean {
    return order.status === OrderStatus.Pending;
}

export function getOrderStatusLabel(status: OrderStatus): string {
    const labels: Record<OrderStatus, string> = {
        [OrderStatus.Pending]: "Pending",
        [OrderStatus.Confirmed]: "Confirmed",
        [OrderStatus.Shipped]: "Shipped",
        [OrderStatus.Delivered]: "Delivered",
        [OrderStatus.Cancelled]: "Cancelled",
    };
    return labels[status];
}

File: tests/fixtures/benchmark_project/src/models/order.py
"""Order domain model."""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum

from src.models.product import Product
from src.models.user import User


class OrderStatus(Enum):
    PENDING = "pending"
    CONFIRMED = "confirmed"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"


@dataclass
class OrderItem:
    product: Product
    quantity: int

    def subtotal(self) -> float:
        return round(self.product.price * self.quantity, 2)


@dataclass
class Order:
    id: int
    user: User
    items: list[OrderItem] = field(default_factory=list)
    status: OrderStatus = OrderStatus.PENDING
    created_at: datetime = field(default_factory=datetime.now)
    notes: str | None = None

    def total(self) -> float:
        return round(sum(item.subtotal() for item in self.items), 2)

    def add_item(self, product: Product, quantity: int = 1) -> None:
        for item in self.items:
            if item.product.id == product.id:
                item.quantity += quantity
                return
        self.items.append(OrderItem(product=product, quantity=quantity))

    def remove_item(self, product_id: int) -> bool:
        for i, item in enumerate(self.items):
            if item.product.id == product_id:
                self.items.pop(i)
                return True
        return False

    def confirm(self) -> None:
        if self.status != OrderStatus.PENDING:
            raise ValueError(f"Cannot confirm order with status {self.status}")
        self.status = OrderStatus.CONFIRMED

    def cancel(self) -> None:
        if self.status in (OrderStatus.SHIPPED, OrderStatus.DELIVERED):
            raise ValueError("Cannot cancel shipped or delivered order")
        self.status = OrderStatus.CANCELLED

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "user": self.user.to_dict(),
            "items": [
                {"product": i.product.to_dict(), "quantity": i.quantity, "subtotal": i.subtotal()}
                for i in self.items
            ],
            "total": self.total(),
            "status": self.status.value,
            "created_at": self.created_at.isoformat(),
        }
